### ISSUE TITLE
segfault found in dev build

### ISSUE THREAD

**Issue by tchaikov** [scylladb/seastar#1824]
this is spotted in our CI. and is found on all 4 combinations of {clang-15, gcc-12} and c++standard 17, c++standard 20
```
          4 - Seastar.unit.alien (Failed)
         16 - Seastar.unit.directory (Failed)
         31 - Seastar.unit.httpd (Failed)
         51 - Seastar.unit.smp (Failed)
         58 - Seastar.unit.thread_context_switch (Failed)
         59 - Seastar.unit.timer (Failed)
         73 - Seastar.app.memcached.memcached (Failed)
```
```
[0/1] cd /home/circleci/project/build/dev/tests/unit && /usr/bin/cmake -E env ASAN_OPTIONS=disable_coredump=0:abort_on_error=1:detect_stack_use_after_return=1:verify_asan_link_order=0 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1 BOOST_TEST_CATCH_SYSTEM_ERRORS=no /home/circleci/project/build/dev/tests/unit/timer_test -c 2
INFO  2023-09-11 12:56:29,828 seastar - Reactor backend: linux-aio
Segmentation fault on shard 0.
Backtrace:
  /home/circleci/project/build/dev/libseastar.so+0x30b2dc
  /home/circleci/project/build/dev/libseastar.so+0x3375d6
  /lib/x86_64-linux-gnu/libc.so.6+0x3bcef
  /home/circleci/project/build/dev/libseastar.so+0x286a5f
  /home/circleci/project/build/dev/libseastar.so+0x272000
  /home/circleci/project/build/dev/libseastar.so+0x267f19
  /home/circleci/project/build/dev/libseastar.so+0x267d63
  /home/circleci/project/build/dev/libseastar.so+0x488972
  /home/circleci/project/build/dev/libseastar.so+0x267c69
  /home/circleci/project/build/dev/libseastar.so+0x308eff
  /home/circleci/project/build/dev/libseastar.so+0x308100
  /home/circleci/project/build/dev/libseastar.so+0x3236ad
  /home/circleci/project/build/dev/libseastar.so+0x3255d8
  /home/circleci/project/build/dev/libseastar.so+0x28f584
  0x590a
  /lib/x86_64-linux-gnu/libc.so.6+0x2350f
  /lib/x86_64-linux-gnu/libc.so.6+0x235c8
  0x5404
Segmentation fault
FAILED: tests/unit/CMakeFiles/test_unit_timer_run /home/circleci/project/build/dev/tests/unit/CMakeFiles/test_unit_timer_run 
```

**Comment by tchaikov** [comment]
not reproduciable locally with dev build, clang-16 on f38.

**Comment by a16bitsysop** [comment]
Running the the aarch64 tests in docker gives lots of these errors:
```
seastar-a9fe22923a412d7d9d348d5d1d1e4c2767b4bb97/build/tests/unit/alien_test -c 2
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not setup Async I/O: 1048576. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr (Function not implemented). Try increasing that number or reducing the amount of logical CPUs available for your application
qemu: uncaught target signal 6 (Aborted) - core dumped
```

Different places but same error, attached log
[aarch64-LastTest.log](https://github.com/scylladb/seastar/files/12588821/aarch64-LastTest.log)

**Comment by tchaikov** [comment]
@a16bitsysop this looks like a different issue. could you please create another issue for tracking this, so we can continue the discussion out there?

**Comment by tchaikov** [comment]
> Running the the aarch64 tests in docker gives lots of these errors:
> 
> ```
> seastar-a9fe22923a412d7d9d348d5d1d1e4c2767b4bb97/build/tests/unit/alien_test -c 2
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Could not setup Async I/O: 1048576. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr (Function not implemented). Try increasing that number or reducing the amount of logical CPUs available for your application
> qemu: uncaught target signal 6 (Aborted) - core dumped
> ```
> 
> Different places but same error, attached log [aarch64-LastTest.log](https://github.com/scylladb/seastar/files/12588821/aarch64-LastTest.log)

issue tracked by #1826

**Comment by tchaikov** [comment]
turns out this is correlated with the protobuf intergration. but the exact root cause is yet known.

**Comment by mykaul** [comment]
> turns out this is correlated with the protobuf intergration. but the exact root cause is yet known.

@tchaikov - it only happens with protobuf integration PR?

**Comment by tchaikov** [comment]
@mykaul correct. FWIW i also tested with the same docker image we use for running CI, but failed to reproduce the test failure after applying the protobuf patch at https://github.com/scylladb/seastar/pull/1818 .

**Comment by tchaikov** [comment]
just managed to get the backtrace. it's related to the kernel used by the host running docker. and turns out it's fixed by #1974. the backtrace is like:

```
Backtrace:
[Backtrace #0]
/home/circleci/project/build/dev/libseastar.so: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=a34550616f1c97e958f47feadd0391f4090578e3, not stripped

void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ??:?
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at reactor.cc:?
seastar::print_with_backtrace(char const*, bool) at reactor.cc:?
seastar::install_oneshot_signal_handler<11, &seastar::sigsegv_action>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) at reactor.cc:?
/lib/x86_64-linux-gnu/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=69389d485a9793dbe873f0ea2c93e02efaa9aa3d, for GNU/Linux 3.2.0, stripp
ed

__dcigettext at ./intl/dcigettext.c:762
void fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >::format_custom_arg<seastar::basic_sstring<char, unsigned int, 15u, true>, fmt::v9::formatter<seastar::basic_sstring<char, unsigned int, 15u, true>, char,
 void> >(void*, fmt::v9::basic_format_parse_context<char, fmt::v9::detail::error_handler>&, fmt::v9::basic_format_context<fmt::v9::appender, char>&) at ??:?
char const* fmt::v9::detail::parse_replacement_field<char, fmt::v9::detail::vformat_to<char>(fmt::v9::detail::buffer<char>&, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<std::conditional<std::
is_same<fmt::v9::type_identity<char>::type, char>::value, fmt::v9::appender, std::back_insert_iterator<fmt::v9::detail::buffer<fmt::v9::type_identity<char>::type> > >::type, fmt::v9::type_identity<char>::type> >, fmt::v9::detail::locale_r
ef)::format_handler&>(char const*, char const*, fmt::v9::detail::vformat_to<char>(fmt::v9::detail::buffer<char>&, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<std::conditional<std::is_same<fmt
::v9::type_identity<char>::type, char>::value, fmt::v9::appender, std::back_insert_iterator<fmt::v9::detail::buffer<fmt::v9::type_identity<char>::type> > >::type, fmt::v9::type_identity<char>::type> >, fmt::v9::detail::locale_ref)::format
_handler&) at ??:?
void fmt::v9::detail::vformat_to<char>(fmt::v9::detail::buffer<char>&, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<std::conditional<std::is_same<fmt::v9::type_identity<char>::type, char>::val
ue, fmt::v9::appender, std::back_insert_iterator<fmt::v9::detail::buffer<fmt::v9::type_identity<char>::type> > >::type, fmt::v9::type_identity<char>::type> >, fmt::v9::detail::locale_ref) at ??:?
seastar::internal::log_buf::inserter_iterator fmt::v9::vformat_to<seastar::internal::log_buf::inserter_iterator, 0>(seastar::internal::log_buf::inserter_iterator, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic
_format_context<fmt::v9::appender, char> >) at ??:?
seastar::logger::do_log(seastar::log_level, seastar::logger::log_writer&) at ??:?
void seastar::logger::log<>(seastar::log_level, seastar::logger::format_info) at ??:?
seastar::internal::make_cpu_stall_detector(seastar::internal::cpu_stall_detector_config) at ??:?
seastar::reactor::reactor(std::shared_ptr<seastar::smp>, seastar::alien::instance&, unsigned int, seastar::reactor_backend_selector, seastar::reactor_config) at ??:?
seastar::smp::allocate_reactor(unsigned int, seastar::reactor_backend_selector, seastar::reactor_config) at ??:?
seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&) at ??:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ??:?
main at ??:?
?? ??:0
?? ??:0
_start at ??:?
```

**Comment by mykaul** [comment]
> just managed to get the backtrace. it's related to the kernel used by the host running docker. and turns out it's fixed by #1974. the backtrace is like:

Great catch!
